[DI] Auto register extension configuration classes as a resource

ro0NL opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #... <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!--highly recommended for new features-->

Auto-register an extension configuration class as a resource from a compiler pass; not implicitly by the base extension class.

Causing some extensions to register its configuration, whereas others dont (e.g. framework bundle).

Fixes consistent cache invalidation whenever a configuration definition changes.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Extension/Extension.php:
```diff
@@ -82,13 +81,8 @@ public function getConfiguration(array $config, ContainerBuilder $container)
         $class = $namespace.'\\Configuration';
-        if (class_exists($class)) {
-            $r = new \ReflectionClass($class);
-            $container->addResource(new FileResource($r->getFileName()));
```
does it hurt to keep this one so that people not using MergeExtensionConfigurationPass don't experience a breakage?

ro0NL replied:
It hurts SF by default :) in terms of registering a duplicate resource. Imo. it just dont belong here..

nicolas-grekas replied:
breakage >>> duplicate resource :)

ro0NL replied:
Understand.. but it's a breakage in behavior.. ie. it doesnt qualify a technical "BC Break".

I'd prefer master+changelog on those, rather then keeping quirky code. Trying to do all.

sstok replied:
Note that duplicate resources are not actually stored.

```php
    public function getResources()
    {
        return array_unique($this->resources);
    }
```

ro0NL replied:
Fair enough. I still think the "change" is considerable, but we can keep things as is here. Doing things twice by default.

ro0NL committed:
[DI] Auto register extension configuration classes as a resource

ro0NL commented:
Updated to 2.7, reverted base `Extension`.

However, we could update built in bundles to bypass `Extension::getConfiguration` right? Initializing the configuration class explicitly per extension (like the framework bundle already does).

That should improve something.

nicolas-grekas commented:
This should be considered as a bug, isn't it? What about adding a test case?

xabbuh commented on src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php:
```diff
@@ -47,6 +48,9 @@ public function process(ContainerBuilder $container)
             $tmpContainer->setResourceTracking($container->isTrackingResources());
             $tmpContainer->addObjectResource($extension);
+            if ($extension instanceof ConfigurationExtensionInterface && null !== $configuration = $extension->getConfiguration($config, $tmpContainer)) {
+                $tmpContainer->addObjectResource($configuration);
```
Now, we have two resources being registered when `getConfiguration()` is used.

ro0NL replied:
Correct :) see https://github.com/symfony/symfony/pull/21057#discussion_r93932858

I first targeted master, doing it only here (with in mind we do a changelog on behavioral change). 

However now we target 2.7, in order to fix the behavior in lower branches. Indeed giving multiple resources by default. But's not really a big a deal, only weird looking at it :)

xabbuh replied:
Hm yeah, thinking about this again two resources shouldn't really hurt.

ro0NL commented:
Test added.

What about core bundles with different implementations regarding `getConfiguration`?

ro0NL commented:
:+1: for keeping thing as is from here.

nicolas-grekas reviewed (approved):
:+1:

fabpot commented:
Thank you @ro0NL.
